osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs

```diff
@@ -536,7 +543,7 @@ public LegacyHitSampleInfo(string name, string? bank = null, int volume = 0, int
             }
 
             public sealed override HitSampleInfo With(Optional<string> newName = default, Optional<string> newBank = default, Optional<string?> newSuffix = default, Optional<int> newVolume = default)
-                => With(newName, newBank, newVolume);
+                => new LegacyHitSampleInfo(newName.GetOr(Name), newBank.GetOr(Bank), newVolume.GetOr(Volume), newSuffix.GetOr(Suffix), IsLayered);
```

Comment:
This change is causing test failures and rightly so because there are several issues with doing this.

This overload is heavily used by several other places in the code (not entirely sure why, looks like something to do with the overload resolution algorithm, probably because it has less default arguments or something). Why is this a problem? Say that you have a `LegacyHitSampleInfo` with `CustomSampleBank = 2`, for instance. Calling this `With()` overload chooses the ctor with `suffix` in it and _not_ the one with `customSampleBank` in it. And the ctor with `suffix` in it, contrary to the one with `customSampleBank` in it, does not set the new `LegacyHitSampleInfo`'s `CustomSampleBank`. This is depended on in a few places (`LegacyBeatmapSkin` first and foremost) and breaks behaviour.

This could be _possibly_ fixable by something along the lines of

```diff
diff --git a/osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs b/osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs
index ec9efcd7e3..e4fd1b106d 100644
--- a/osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs
+++ b/osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs
@@ -543,7 +543,11 @@ public LegacyHitSampleInfo(string name, string? bank = null, int volume = 0, int
             }
 
             public sealed override HitSampleInfo With(Optional<string> newName = default, Optional<string> newBank = default, Optional<string?> newSuffix = default, Optional<int> newVolume = default)
-                => new LegacyHitSampleInfo(newName.GetOr(Name), newBank.GetOr(Bank), newVolume.GetOr(Volume), newSuffix.GetOr(Suffix), IsLayered);
+            {
+                return newSuffix.HasValue
+                    ? new LegacyHitSampleInfo(newName.GetOr(Name), newBank.GetOr(Bank), newVolume.GetOr(Volume), newSuffix.GetOr(Suffix), IsLayered)
+                    : new LegacyHitSampleInfo(newName.GetOr(Name), newBank.GetOr(Bank), newVolume.GetOr(Volume), CustomSampleBank, IsLayered);
+            }
 
             public virtual LegacyHitSampleInfo With(Optional<string> newName = default, Optional<string> newBank = default, Optional<int> newVolume = default,
                                                     Optional<int> newCustomSampleBank = default,

```

except with more cases, but problem number 2 is `FileHitSampleInfo` which derives from `LegacyHitSampleInfo`. The way this overriding and double-pathery works is fundamentally incompatible with inheritance and I don't know how you fix it without a rewrite.

With all of that in mind - what was the point of the `Suffix` thing and the entire `CustomSampleBank` / `Suffix` dual ctor path thing? What was it trying to achieve? Because if I'm honest I find it weird and confusing and would much rather not have it exist.

Probably not much I can review further here until we get that straightened out.

Reply:
I'm not really sure it needs to exist, and I hope it doesn't because I share the same feelings. The inheritance is a mess.

Maybe it's to make saving work? Will require further investigation.

Reply:
I misread your question, so here's a proper answer:

> With all of that in mind - what was the point of the Suffix thing and the entire CustomSampleBank / Suffix dual ctor path thing? What was it trying to achieve? Because if I'm honest I find it weird and confusing and would much rather not have it exist.

Basically, without it the suffix gets ignored. See the original code which accepts a suffix in, but does not pass it through.

But given the feedback from @nekodex, this may have been a red herring as the suffix I was applying is much later in the process, so I think it's probably fine to revert this change.

Reply:
Yeah, I've confirmed this change isn't required, so will revert immediately.